src/javascript/hooks.tsx

```diff
@@ -0,0 +1,117 @@
+export const useEventValue = (setValue: (value: string) => void) =>
+  React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) =>
+      setValue(event.target.value),
+    [setValue]
+  );
+
+// Convert a callback taking a boolean into a callback taking event.target.checked
+export const useEventChecked = (setChecked: (checked: boolean) => void) =>
+  React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) =>
+      setChecked(event.target.checked),
+    [setChecked]
+  );
+
+// Same as `useState`, but also store the value in `localStorage` with a
+// `key` whenever it is updated, and initialize the state from `localStorage`,
+// if present. If `subscribe` is true, also propogate changes from localStorage
+// from other tabs into the state.
+export const useLocalStorage = function(
+  key: string,
+  initialValue: string | (() => string),
+  subscribe: boolean = true
+): [string, React.Dispatch<React.SetStateAction<string>>] {
+  // Set up the state; initialize from localStorage if available.
+  const [currentValue, setValue] = React.useState(() => {
+    const storedValue = window.localStorage.getItem(key);
+    return storedValue !== null
+      ? storedValue
+      : initialValue instanceof Function
+      ? initialValue()
+      : initialValue;
+  });
+
+  // Create a wrapper around setValue that also stores to localStorage
+  const setValueAndStore = React.useCallback(
+    (value: React.SetStateAction<string>) =>
+      setValue((oldValue: string) => {
+        let newValue = value instanceof Function ? value(oldValue) : value;
+        window.localStorage.setItem(key, newValue);
+        return newValue;
+      }),
+    [setValue, key]
+  );
+
+  // If subscribe is true, subscribe to localStorage events and call
+  // setValue when there are changes
+  React.useEffect(() => {
+    if (subscribe) {
+      const listener = (event: StorageEvent) => {
+        if (event.storageArea === window.localStorage) {
+          if (event.key === null) {
+            // null key key means that there was a global storage clear event
+            setValue(
+              initialValue instanceof Function ? initialValue() : initialValue
+            );
+          } else if (event.key === key) {
+            if (event.newValue === null) {
+              setValue(
+                initialValue instanceof Function ? initialValue() : initialValue
+              );
+            } else {
+              setValue(event.newValue);
+            }
+          }
+        }
+      };
+
+      window.addEventListener("storage", listener);
+
+      return () => window.removeEventListener("storage", listener);
+    }
+  }, [key, setValue, initialValue, subscribe]);
+
+  return [currentValue, setValueAndStore];
+};
+
+// Same as `useLocalStorage`, but typed. Uses `JSON.parse` and
+// `JSON.stringify`` to convert values back and forth between strings
+// in `localStorage`.
+export function useTypedLocalStorage<T>(
```

Comment:
Since this only works with types that are JSON.stringify-able, you might want to see if you can narrow the types away from T to something more specific. Off the top of my head, I'm not sure if there's a clean way to do this, but I know this wouldn't work with `Map<NotString, Value>`.

Reply:
I was trying that, but you very quickly run into a wall where `JSON.parse` returns `any`.